src/preload.py

```diff
@@ -42,116 +45,371 @@
+    if size is not None:
+        args.extend(["--sizelimit", size])
+    args.append(image)
+    device = losetup(*args).stdout.decode("utf8").strip()
+    yield device
+    losetup("-d", device)
 
 
 @contextmanager
-def mount_context_manager(partition_name, extra_options="", image=None):
-    mountpoint = mount_partition(partition_name, extra_options, image)
+def device_mount_context_manager(device):
+    mountpoint = mkdtemp()
+    mount(device, mountpoint)
     yield mountpoint
     umount(mountpoint)
     os.rmdir(mountpoint)
 
 
-def losetup_partition(partition_name, image=None):
-    part = get_partition(partition_name, image)
-    offset, size = get_offset_and_size(part)
-    return losetup(
-        "-o",
-        offset,
-        "--sizelimit",
-        size,
-        "-P",
-        "--show",
-        "-f",
-        part["image"],
-    ).stdout.decode("utf8").strip()
+@contextmanager
+def mount_context_manager(image, offset=None, size=None):
+    with losetup_context_manager(image, offset, size) as device:
+        with device_mount_context_manager(device) as mountpoint:
+            yield mountpoint
+
+
+class FilePartition(object):
+    def __init__(self, image, label):
+        self.image = image
+        self.label = label
+
+    def losetup_context_manager(self):
+        return losetup_context_manager(self.image)
+
+    def mount_context_manager(self):
+        return mount_context_manager(self.image)
+
+    def resize(self, additional_bytes):
+        expand_file(self.image, additional_bytes)
+        expand_filesystem(self)
+
+    def str(self):
+        return self.image
+
+
+class Partition(object):
+    def __init__(
+        self,
+        partition_table,
+        number,
+        node=None,
+        start=None,
+        size=None,
+        type=None,
+        uuid=None,
+        name=None,
+        bootable=False,
+    ):
+        self.partition_table = partition_table
+        self.number = number
+        self.parent = None
+        self.node = node
+        self.start = start
+        self.size = size
+        self.type = type
+        self.uuid = uuid
+        self.name = name
+        self.bootable = bootable
+        # label, not part of the sfdisk script
+        self.label = self._get_label()
+
+    def _get_label(self):
+        with self.losetup_context_manager() as device:
+            out = file("-s", device).stdout.decode("utf8").strip()
+            # "label:" is for fat partitions,
+            # "volume name" is for ext partitions
+            # "BTRFS Filesystem label" is for btrfs partitions
+            match = search(
+                '(BTRFS Filesystem label|label:|volume name) "(.*)"',
+                out,
+            )
+            if match is not None:
+                return match.groups()[1].strip()
+
+    def set_parent(self, parent):
+        # For logical partitions on MBR disks we store the parent extended
+        # partition
+        assert self.partition_table.label == "dos"
```

Comment:
Do we want to leave an assert in production code?

Reply:
Yes, preload.py has several asserts.
I think it might help debugging.
Without asserts, weird errors will happen later.

Reply:
Gotcha, just wanted to make sure it was being considered.